r/R/dplyr-functions.R

```diff
@@ -57,6 +57,30 @@ nse_funcs$cast <- function(x, target_type, safe = TRUE, ...) {
   Expression$create("cast", x, options = opts)
 }
 
+nse_funcs$coalesce <- function(...) {
+  if (missing(..1)) {
+    abort("At least one argument must be supplied to coalesce()")
+  }
+  # TODO: if an option is added to the coalesce kernel to treat NaN as NA,
+  # use that to simplify the code here (ARROW-13389)
+  args <- lapply(list2(...), function(arg) {
+    if (!inherits(arg, "Expression")) {
+      arg <- Expression$scalar(arg)
+    }
+    if (arg$type_id() %in% TYPES_WITH_NAN) {
+      # replace NaN with NA, using Arrow's smallest float type to avoid casting
+      # smaller float types to larger float types
```

Comment:
This approach is flawed because it replaces `NaN` with `NA` in the result when all the values are `NaN`

Reply:
Oh, I think the solution is to skip the last iteration of the `lapply`

Reply:
Fixed in 3a91ee9

Reply:
FYI, when all values in the row are `NA` or `NaN`, `dplyr::coalesce()` returns the _last_ one, which is why this solution works:
```
> dplyr::coalesce(NA, NaN)
[1] NaN
> dplyr::coalesce(NaN, NA)
[1] NA
```